### PR TITLE
Set log level from parameter or envvar.

### PR DESCRIPTION
**Pull request by dbrumann**
This PR is yet another attempt at solving issue #220, just like #245 and #289 before.

Sorry, for opening yet another PR for this, but since there wasn't much activity I figured I give it a go.

### CODE REVIEW

**Comment by SerkanYildiz** [comment]
Hi @dbrumann, we're in the phase of extending the Extension class of Monolog to have this functionality.
However we would like to have this in the bundle, so that we don't manage it ourselves.

Can we help you to get this functionality merged in the bundle itself?

Thanks in advance!

**Comment by dbrumann** [comment]
Thanks @SerkanYildiz. The failing tests should be addressed by #360. I hope after merging that we can also merge this. I guess other than reviewing if this is the appropriate way to solve this, there is not much else to do here, right now.

File: DependencyInjection/MonologExtension.php [hunk 1]
```diff
@@ -41,9 +42,31 @@ class MonologExtension extends Extension
 
     private $swiftMailerHandlers = [];
 
-    private function levelToMonologConst($level)
+    private function levelToMonologConst($level, ContainerBuilder $container)
     {
-        return is_int($level) ? $level : constant('Monolog\Logger::'.strtoupper($level));
+        if ($level === null || is_int($level)) {
```
Comment (lyrixx) [inline DependencyInjection/MonologExtension.php:47]: Why do you allow returning  `null`?

File: Tests/DependencyInjection/MonologExtensionTest.php [hunk 2]
```diff
@@ -611,10 +612,83 @@ public function v1AddedDataProvider()
...
+        $this->expectException(InvalidArgumentException::class);
+        $this->expectExceptionMessage('Could not match "%some_param%" to a log level.');
+
+        $loader->load($config, $container);
+    }
+
+    /**
+     * @dataProvider provideLoglevelParameterConfig
+     */
+    public function testLogLevelfromParameter($parameters, $config, $expectedClass, $expectedArgs)
```
Comment (lyrixx) [inline Tests/DependencyInjection/MonologExtensionTest.php:630]:
```suggestion
    public function testLogLevelfromParameter(array $parameters, array $config, string $expectedClass, array $expectedArgs)
```

File: Tests/DependencyInjection/MonologExtensionTest.php [hunk 3]
```diff
@@ -611,10 +612,83 @@ public function v1AddedDataProvider()
...
+        foreach ($parameters as $name => $value) {
+            $container->setParameter($name, $value);
+        }
+        $loader = new MonologExtension();
+        $config = [['handlers' => ['main' => $config]]];
+        $loader->load($config, $container);
+
+        $definition = $container->getDefinition('monolog.handler.main');
+        $this->assertDICDefinitionClass($definition, $expectedClass);
+
```
Comment (lyrixx) [inline Tests/DependencyInjection/MonologExtensionTest.php:642]:
```suggestion
```

**Review by lyrixx — COMMENTED** [review]
Looks good, but I left some comments. 
Thanks for working on this

File: DependencyInjection/MonologExtension.php (hunk 1 above)
Comment (dbrumann) [reply DependencyInjection/MonologExtension.php:47]: I figured not specifying a level will not pass it and cause the handler's default to be used. I will revert it

**Comment by lyrixx** [comment]
Could you add a note in the CHANGELOG? after that I'm 👍🏼

**Review by lyrixx — APPROVED** [review]

**Comment by dbrumann** [comment]
I just tried this out in a project and noticed an issue. The following config (when no env var is set) will not work:

```yaml
# config/packages/monolog/dev.yaml
parameters:
    env(LOG_LEVEL): ~
monolog:
    handlers:
        main:
            type: stream
            path: "%kernel.logs_dir%/%kernel.environment%.log"
            level: '%env(LOG_LEVEL)%'
```

I am working on a fix for this.

**Review by lyrixx — APPROVED** [review]
OK, let's go 👍🏼

**Comment by lyrixx** [comment]
It took time, but here we go, this is in now. Thank you very much @dbrumann.

**Comment by eexit** [comment]
Hello,

Quick question: is it possible to combine this env var with the default resolver?
I'm trying to do this but I'm getting the following error: `'%env(default:info:LOG_LEVEL)%'`.

```
In MergeExtensionConfigurationPass.php line 217:
!!
!!    Using a cast in "env(default:info:LOG_LEVEL)" is incompatible with resoluti
!!    on at compile time in "Symfony\Bundle\MonologBundle\DependencyInjection\Mon
!!    ologExtension". The logic in the extension should be moved to a compiler pa
!!    ss, or an env parameter with no cast should be used instead.
```

Any insight?
Thanks!

**Comment by dbrumann** [comment]
No, unfortunately not.  This ties into the issue reported in #413 The log level needs to be evaluated during container compilation to be able to build the logger.

As far as I can tell we can only get the value and not use resolvers at that point. I am not 100% sure if this is really a general limitation or caused by how I implemented it. It might also have changed - I haven't really kept up to date with the changes around resolving env vars. So it might be doable, but I am afraid I don't have much time to check it at the moment. Feel free to have a deeper look, if you can.

**Comment by eexit** [comment]
Ok, thanks @dbrumann!

**Pull request merged**